tests/admin_scripts/tests.py

```diff
@@ -1867,6 +1868,20 @@ def _test(self, args, option_b="'2'"):
         )
 
 
+class ExecuteFromCommandLine(SimpleTestCase):
+    def test_program_name_from_argv(self):
+        """
+        Program name is computed from the execute_from_command_line()'s argv
+        argument, not sys.argv.
+        """
+        args = ['help', 'shell']
+        with captured_stdout() as out, captured_stderr() as err:
+            with mock.patch('sys.argv', [None] + args):
+                execute_from_command_line(['django-admin'] + args)
+        self.assertIn('usage: django-admin shell', out.getvalue())
```

Comment:
This is fine as long as it's clear that the patch to `execute_from_command_line` in this PR isn't actually the thing that makes the usage message say "django-admin" rather than "None". (As far as I can tell, there's no way to induce the modified `CommandParser` to issue its usage message.)